SignalServiceKit/src/Util/ParamParser.swift

```diff
@@ -0,0 +1,126 @@
+//
+//  Copyright (c) 2018 Open Whisper Systems. All rights reserved.
+//
+
+import Foundation
+
+// A DSL for parsing expected and optional values from a Dictionary, appropriate for
+// validating a service response.
+//
+// Additionally it includes some helpers to DRY up common conversions.
+//
+// Rather than exhaustively enumerate accessors for types like `requireUInt32`, `requireInt64`, etc.
+// We instead leverage generics at the call site.
+//
+//     do {
+//         // Required
+//         let name: String = try paramParser.require(key: "name")
+//         let count: UInt32 = try paramParser.require(key: "count")
+//
+//         // Optional
+//         let last_seen: Date? = try paramParser.allow(key: "last_seen")
```

Comment:
I find "require/allow" unintuitive, especially "allow".  Personally, I'd prefer "parseRequired/parseOptional" or "required/optional".  Strictly TIOLI

Reply:
Since a primary goal of this is to allow concise calling, I chose pretty terse names, but I can live with `required`/`optional`, since I think "parse" is somewhat superfluous given it's a Parser class with two method flavors.